Feature/userspace interface

SAreebZaheer opened the pull request.

SAreebZaheer committed:
feat(userspace): add libmutex public API header

Add comprehensive public API header for MUTEX userspace library.
Defines high-level C API for interacting with kernel proxy via file descriptors.

Functions:
- mprox_create() - Create proxy fd (syscall wrapper)
- mprox_enable/disable() - Control proxy state
- mprox_set_config/get_config() - Proxy configuration
- mprox_set_socks5/http() - Convenience config functions
- mprox_get_stats() - Statistics retrieval
- mprox_wait_event() - Event notification support
- Utility functions for version, error handling

Part of Branch 3: userspace-interface

SAreebZaheer committed:
feat(userspace): implement libmutex library

Implement complete userspace library for MUTEX proxy API:
- Wrapper for mprox_create() syscall (syscall 471)
- ioctl-based control functions (enable/disable)
- Configuration functions with validation
- IPv4/IPv6 address parsing for SOCKS5/HTTP proxies
- Statistics retrieval via ioctl
- poll() support for event notification
- Error handling and errno mapping

All functions perform input validation and proper error handling.
Part of Branch 3: userspace-interface

SAreebZaheer committed:
docs(userspace): add comprehensive API documentation

Add detailed API reference for libmutex:
- Function prototypes and descriptions
- Parameter documentation
- Return values and error codes
- Data structure definitions
- Usage examples and best practices
- Thread safety guidelines
- Error handling patterns

300+ lines of documentation with code examples.
Part of Branch 3: userspace-interface

SAreebZaheer committed:
build(userspace): add Makefile for libmutex

Add build system for MUTEX userspace library:
- Build shared library (libmutex.so.0.1.0)
- Build static library (libmutex.a)
- Generate pkg-config file
- Install/uninstall targets
- Support for custom PREFIX
- SONAME versioning

Generates both shared and static libraries with proper versioning.
Part of Branch 3: userspace-interface

SAreebZaheer committed:
feat(userspace): add mprox CLI tool

Add command-line utility for MUTEX proxy management:

Commands:
- create: Create new proxy file descriptor
- enable/disable: Control proxy state
- config: Configure proxy settings
- status: Show proxy status
- stats: Show statistics

Options:
- Address, port, type (socks5/http/https)
- Flags: global, cloexec, nonblock
- Verbose output

Uses libmutex for all operations.
Part of Branch 3: userspace-interface

SAreebZaheer committed:
build(userspace): add Makefile for mprox CLI

Add build system for mprox command-line tool:
- Links against libmutex
- Install to PREFIX/bin
- Uninstall target
- Basic test target

Part of Branch 3: userspace-interface

SAreebZaheer committed:
feat(userspace): add example programs

Add three example programs demonstrating MUTEX proxy API:

1. simple_proxy.c:
   - Basic usage walkthrough
   - Create fd, configure SOCKS5, enable proxy
   - Make HTTP connection through proxy
   - View statistics

2. multi_fd.c:
   - Multiple independent proxy fds
   - Different configurations per fd
   - Demonstrate fd lifecycle management

3. poll_example.c:
   - Event-driven programming with poll()
   - Non-blocking mode
   - Event notification and statistics monitoring

Part of Branch 3: userspace-interface

SAreebZaheer committed:
docs(userspace): add examples documentation and build system

Add README for example programs with:
- Description of each example
- Build instructions
- Usage examples
- Testing without proxy server

Add Makefile to build all examples.
Part of Branch 3: userspace-interface

SAreebZaheer committed:
build(userspace): add top-level build system and documentation

Add master Makefile for all userspace components:
- Builds library, CLI, and examples
- Cascading clean target
- Install/uninstall for library and CLI
- Help target with documentation

Add comprehensive README covering:
- Architecture overview
- Building and installation
- Usage examples for library and CLI
- API documentation references
- Troubleshooting guide
- Development guidelines

Part of Branch 3: userspace-interface

SAreebZaheer committed:
docs: add Branch 3 implementation summary

Add comprehensive summary document for userspace-interface branch:
- Deliverables checklist
- Architecture explanation
- API functions implemented
- Testing instructions
- Integration points with other branches
- Known limitations
- Verification checklist

Part of Branch 3: userspace-interface

SAreebZaheer committed:
docs: update project README for Branch 3 completion

Update project documentation to reflect Branch 3 completion:
- Note userspace library implementation
- Add build instructions for userspace components
- Update project structure

Part of Branch 3: userspace-interface

SAreebZaheer committed:
chore(module): update Makefile

Update module Makefile for compatibility with userspace build.
Part of Branch 3: userspace-interface

SAreebZaheer committed:
docs: update main README for Branch 3 completion

Major updates to reflect userspace interface implementation:
- Update current status to show Branch 3 complete
- Add userspace components to project structure
- Update build instructions with userspace examples
- Add userspace API quick start section
- Update features list with fd-based interface
- Add documentation links for libmutex API
- Update version to 0.3.0
- Update milestone completion status

Branch 3 (userspace-interface) is now complete with full
library, CLI tool, and example programs.

SAreebZaheer committed:
docs: remove testing section from README

Remove redundant testing section that duplicated information
already available in TESTING.md and test scripts.

Testing details are better maintained in:
- docs/TESTING.md for comprehensive test documentation
- src/module/test_module.sh for automated tests
- Individual component READMEs for specific testing

Keeps main README focused on quick start and overview.

Hamza-Bin-Aamir commented on src/module/Makefile:
```diff
@@ -15,7 +15,7 @@ KDIR := /lib/modules/$(shell uname -r)/build
 
 # Use system's default gcc
-export CC := gcc
+export CC := x86_64-linux-gnu-gcc-12
```
interesting, why force a specific compiler?

SAreebZaheer replied:
wasnt building until it was specified. will look to see a better approach to this

Hamza-Bin-Aamir replied:
maybe add it as a cli argument so by default it uses gcc and if that doesnt work we can specify exact compiler using cli argument?

SAreebZaheer replied:
yep yep will do, good option

SAreebZaheer replied:
done

Hamza-Bin-Aamir commented on src/userspace/examples/README.md:
````diff
@@ -0,0 +1,105 @@
+
+```bash
+# Using dante (sockd)
+sudo apt-get install dante-server
````
pls also add fedora instructions
```suggestion
sudo apt-get install dante-server
# fedora/cent-OS/etc.
sudo dnf install dante-server
```

Hamza-Bin-Aamir reviewed (approved):
Looks good to me! Just check out the suggestions i added

SAreebZaheer committed:
Update src/userspace/examples/README.md

added instructions for fedora

Co-authored-by: Hamza Bin Aamir <148682978+Hamza-Bin-Aamir@users.noreply.github.com>

SAreebZaheer committed:
build(module): make compiler configurable via CC variable

Allow overriding the compiler via command-line argument while
keeping gcc as the default:
- Change CC assignment from := to ?= for override support
- Add export CC to pass compiler to kernel build system
- Update help text with CC option and usage examples

Usage:
  make              # Use default gcc
  make CC=gcc-12    # Use gcc-12
  make CC=clang     # Use clang

This improves build flexibility across different systems and
compiler versions without modifying the Makefile.
